### PR TITLE
Introduce TrajectoryConnectivityState.

### PR DESCRIPTION
**Pull request by cschuet**
This class will be used to track the connectivity state (including the
last connection time) between pairs of trajectories.

### CODE REVIEW

File: cartographer/mapping_3d/sparse_pose_graph.cc [hunk 1]
```diff
@@ -330,8 +330,14 @@ void SparsePoseGraph::HandleWorkQueue() {
         for (const Constraint& constraint : result) {
           CHECK_EQ(constraint.tag,
                    mapping::SparsePoseGraph::Constraint::INTER_SUBMAP);
-          connected_components_.Connect(constraint.node_id.trajectory_id,
-                                           constraint.submap_id.trajectory_id);
+          mapping::NodeId last_submap_node_id =
```
Comment (wohe) [inline cartographer/mapping_3d/sparse_pose_graph.cc]: `const`?

File: cartographer/mapping_3d/sparse_pose_graph.cc [hunk 2]
```diff
@@ -330,8 +330,14 @@ void SparsePoseGraph::HandleWorkQueue() {
         for (const Constraint& constraint : result) {
```
Comment (wohe) [inline cartographer/mapping_3d/sparse_pose_graph.cc]: Instead of a comment: pull this out into a separate function `UpdateTrajectoryConnectivity(const sparse_pose_graph::ConstraintBuilder::Result& result)`?

File: cartographer/mapping/connected_components.cc [hunk 3]
```diff
@@ -91,6 +91,17 @@ std::vector<std::vector<int>> ConnectedComponents::Components() {
   return result;
 }
 
+std::vector<int> ConnectedComponents::GetComponent(int trajectory_id) {
```
Comment (wohe) [inline cartographer/mapping/connected_components.cc]: `const int trajectory_id`.

File: cartographer/mapping/connected_components.cc [hunk 4]
```diff
@@ -91,6 +91,17 @@ std::vector<std::vector<int>> ConnectedComponents::Components() {
...
+  int set_id = FindSet(trajectory_id);
```
Comment (wohe) [inline cartographer/mapping/connected_components.cc]: `const int`

File: cartographer/mapping/connected_components.cc [hunk 5]
```diff
@@ -91,6 +91,17 @@ std::vector<std::vector<int>> ConnectedComponents::Components() {
...
+  for (const auto& entry : forest_) {
+    if (FindSet(entry.first) == set_id) {
+      trajectory_ids.push_back(entry.first);
+    }
+  }
+  return trajectory_ids;
+}
+
 int ConnectedComponents::ConnectionCount(const int trajectory_id_a,
                                             const int trajectory_id_b) {
```
Comment (wohe) [inline cartographer/mapping/connected_components.cc]: Fix indentation?

File: cartographer/mapping/connected_components.h [hunk 6]
```diff
@@ -63,6 +63,10 @@ class ConnectedComponents {
   // The trajectory IDs, grouped by connectivity.
   std::vector<std::vector<int>> Components() EXCLUDES(lock_);
 
+  // The list of trajectory IDs that belong to the same connected component as
+  // 'trajectory_id'.
+  std::vector<int> GetComponent(int trajectory_id) EXCLUDES(lock_);
+
```
Comment (wohe) [inline cartographer/mapping/connected_components.h]: To achieve smaller PRs: This could have easily pulled out into a separate PR.

File: cartographer/mapping/trajectory_connectivity_state.cc [hunk 7]
```diff
@@ -0,0 +1,76 @@
...
+#include "cartographer/mapping/trajectory_connectivity_state.h"
+
+namespace cartographer {
+namespace mapping {
+
+TrajectoryConnectivityState::TrajectoryConnectivityState()
+    : connected_components_(),
+      last_connection_time_map_() {}
+
+void TrajectoryConnectivityState::Add(int trajectory_id) {
```
Comment (wohe) [inline cartographer/mapping/trajectory_connectivity_state.cc]: `const int`. Similar for other arguments in this file that are never changed.

File: cartographer/mapping/trajectory_connectivity_state.cc [hunk 8]
```diff
@@ -0,0 +1,76 @@
...
+ * limitations under the License.
+ */
+
+#include "cartographer/mapping/trajectory_connectivity_state.h"
+
+namespace cartographer {
+namespace mapping {
+
+TrajectoryConnectivityState::TrajectoryConnectivityState()
+    : connected_components_(),
```
Comment (wohe) [inline cartographer/mapping/trajectory_connectivity_state.cc]: Redundant? Remove?

File: cartographer/mapping/trajectory_connectivity_state.cc [hunk 9]
```diff
@@ -0,0 +1,76 @@
...
+  if (TransitivelyConnected(trajectory_id_a, trajectory_id_b)) {
+    // The trajectories are transitively connected, i.e. they belong to the same
+    // connected component. In this case we only update the last connection time
+    // of those two trajectories.
+    auto sorted_pair = std::minmax(trajectory_id_a, trajectory_id_b);
+    if (last_connection_time_map_[sorted_pair] < time) {
+      last_connection_time_map_[sorted_pair] = time;
+    }
+  } else {
+    // The connection between these two trajectories is about to join to
```
Comment (wohe) [inline cartographer/mapping/trajectory_connectivity_state.cc]: Instead of describing what we do here, describe why: we want to quickly change to more efficient loop closure searches when connected components are joined.

File: cartographer/mapping_2d/sparse_pose_graph.cc [hunk 10]
```diff
@@ -180,7 +180,7 @@ void SparsePoseGraph::ComputeConstraint(const mapping::NodeId& node_id,
         submap_id, submap_data_.at(submap_id).submap.get(), node_id,
         trajectory_nodes_.at(node_id).constant_data.get());
   } else {
-    if (connected_components_.TransitivelyConnected(
+    if (trajectory_connectivity_state_.TransitivelyConnected(
```
Comment (wohe) [inline cartographer/mapping_2d/sparse_pose_graph.cc]: Implementing `TrajectoryConnectivityState` and using it could have become two separate PRs. We are not really making use of the new functionality here anyway.

File: cartographer/mapping/trajectory_connectivity_state.h [hunk 11]
```diff
@@ -0,0 +1,80 @@
...
+
+  // Return the last connection count between the two trajectories. If either of
+  // the trajectories is untracked or they have never been connected returns the
+  // beginning of time.
+  common::Time LastConnectionTime(int trajectory_id_a, int trajectory_id_b);
+
+ private:
+  ConnectedComponents connected_components_;
+
+  // Tracks the last time a direction connection between two trajectories has
```
Comment (wohe) [inline cartographer/mapping/trajectory_connectivity_state.h]: typo: direct connection?

File: cartographer/mapping/trajectory_connectivity_state.cc (hunk 8 above)
Comment (cschuet) [reply cartographer/mapping/trajectory_connectivity_state.cc]: Done

File: cartographer/mapping/trajectory_connectivity_state.cc (hunk 7 above)
Comment (cschuet) [reply cartographer/mapping/trajectory_connectivity_state.cc]: Done

File: cartographer/mapping/trajectory_connectivity_state.cc (hunk 9 above)
Comment (cschuet) [reply cartographer/mapping/trajectory_connectivity_state.cc]: Done

File: cartographer/mapping/trajectory_connectivity_state.h (hunk 11 above)
Comment (cschuet) [reply cartographer/mapping/trajectory_connectivity_state.h]: Done

File: cartographer/mapping_2d/sparse_pose_graph.cc (hunk 10 above)
Comment (cschuet) [reply cartographer/mapping_2d/sparse_pose_graph.cc]: Noted for the future.

File: cartographer/mapping_3d/sparse_pose_graph.cc (hunk 2 above)
Comment (cschuet) [reply cartographer/mapping_3d/sparse_pose_graph.cc]: Done

File: cartographer/mapping_3d/sparse_pose_graph.cc (hunk 1 above)
Comment (cschuet) [reply cartographer/mapping_3d/sparse_pose_graph.cc]: Done

File: cartographer/mapping/trajectory_connectivity_state.h [hunk 12]
```diff
@@ -0,0 +1,80 @@
...
+  // Add a trajectory which is initially connected to only itself.
+  void Add(int trajectory_id);
+
+  // Connect two trajectories. If either trajectory is untracked, it will be
+  // tracked. This function is invariant to the order of its arguments. Repeated
+  // calls to Connect increment the connectivity count and update the last
+  // connected time.
+  void Connect(int trajectory_id_a,
+               int trajectory_id_b,
+               const common::Time& time);
```
Comment (wohe) [inline cartographer/mapping/trajectory_connectivity_state.h]: Consistency: `common::Time` is passed by value since it is just an `int64`.

File: cartographer/mapping_2d/sparse_pose_graph.cc [hunk 13]
```diff
@@ -314,8 +314,14 @@ void SparsePoseGraph::HandleWorkQueue() {
...
                    mapping::SparsePoseGraph::Constraint::INTER_SUBMAP);
-          connected_components_.Connect(constraint.node_id.trajectory_id,
-                                           constraint.submap_id.trajectory_id);
+          mapping::NodeId last_submap_node_id =
+              *submap_data_.at(constraint.submap_id).node_ids.rbegin();
+          common::Time time = std::max(
+              trajectory_nodes_.at(constraint.node_id).constant_data->time,
+              trajectory_nodes_.at(last_submap_node_id).constant_data->time);
+          trajectory_connectivity_state_.Connect(constraint.node_id.trajectory_id,
+                                        constraint.submap_id.trajectory_id,
```
Comment (wohe) [inline cartographer/mapping_2d/sparse_pose_graph.cc]: Did you run `clang-format`? Also: Extract a function similar to 3D?
